flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java

```diff
@@ -358,17 +360,51 @@ public static boolean stateBackendFromApplicationOrConfigOrDefaultUseManagedMemo
         return false;
     }
 
+    /**
+     * Load state backend which may wrap the original state backend for recovery.
+     *
+     * @param originalStateBackend StateBackend loaded from application or config.
+     * @param classLoader User code classloader.
+     * @param keyedStateHandles The state handles for restore.
+     * @return Wrapped state backend for recovery.
+     * @throws DynamicCodeLoadingException Thrown if keyed state handles of wrapped state backend
+     *     are found and the class was not found or could not be instantiated.
+     */
+    public static StateBackend loadStateBackendFromKeyedStateHandles(
+            StateBackend originalStateBackend,
+            ClassLoader classLoader,
+            Collection<KeyedStateHandle> keyedStateHandles)
+            throws DynamicCodeLoadingException {
+        if (keyedStateHandles.stream()
+                        .anyMatch(stateHandle -> stateHandle instanceof ChangelogStateBackendHandle)
+                && !isChangelogStateBackend(originalStateBackend)) {
+            return loadChangelogStateBackend(originalStateBackend, classLoader, true);
+        }
+        return originalStateBackend;
+    }
+
+    private static boolean isChangelogStateBackend(StateBackend backend)
+            throws DynamicCodeLoadingException {
+        try {
+            return Class.forName(CHANGELOG_STATE_BACKEND) == backend.getClass();
+        } catch (ClassNotFoundException e) {
```

Comment:
1. Do we really need to instantiate the class here, can't we just check the class name? Or is it because of some classloading issues?
2. What if the Changelog backend is wrapped by some other backend? Should we check `getDelegatedStateBackend` recursively?

Reply:
Agree. We may just need to compare string of their name.
For the second problem, I think it needs more judgement in `loadStateBackendFromKeyedStateHandles` but it may deserve more discussion, and I just added some comments in the method currently.